docs/src/threadpool.rst

```diff
@@ -32,35 +55,139 @@ Data types
-    was cancelled using :c:func:`uv_cancel` `status` will be ``UV_ECANCELED``.
+    thread after the work on the executor has been completed or cancelled.
+    If the work was cancelled using :c:func:`uv_cancel`,
+    then `status` will be ``UV_ECANCELED``.
+
+.. c:type:: uv_executor_t
+
+    Executor type. Use when overriding the default threadpool.
+
+.. c:type:: void (*uv_executor_submit_func)(uv_executor_t* executor, uv_work_t* req, uv_work_options_t* opts)
+
+    Called when the executor should handle this request.
+
+.. c:type:: int (*uv_executor_cancel_func)(uv_executor_t* executor, uv_work_t* req)
 
+    Called when someone wants to cancel a previously-submit'ed request.
+    Return ``UV_EBUSY`` if you cannot cancel it.
+
+.. seealso:: :c:func:`uv_cancel_t`.
+
+.. c:type:: int (*uv_executor_done_cb)(uv_work_t* req)
+
+    libuv sets this during a successful call to :c:func:`uv_replace_executor`.
+    An executor should invoke this CB once finished with a request.
+    This CB is thread safe.
+
+.. c:type:: int uv_replace_executor(uv_executor_t* executor)
+
+    Replace the default libuv executor with this user-defined one.
+    Must be called before any work is submitted to the default libuv executor.
+    Returns 0 on success.
+
+.. c:type:: uv_work_options_t
+
+    Options for guiding the executor in its policy decisions.
 
 Public members
 ^^^^^^^^^^^^^^
 
+.. c:member:: void* uv_work_t.reserved[0]
+
+    Space for arbitrary data. libuv does not use this field.
+    This is intended for use by an executor implementation.
+
 .. c:member:: uv_loop_t* uv_work_t.loop
 
     Loop that started this request and where completion will be reported.
     Readonly.
 
+.. c:member:: uv_work_cb uv_work_t.work_cb
+
+    Executor should invoke this, not on the event loop.
+
 .. seealso:: The :c:type:`uv_req_t` members also apply.
 
+.. c:member:: uv_executor_submit_func uv_executor_t.submit
+
+    Must be non-NULL.
+
+.. c:member:: uv_executor_cancel_func uv_executor_t.cancel
+
+    Can be NULL.
+    If NULL, calls to :c:function:`uv_cancel` will return ``UV_ENOSYS``.
+
+.. c:member:: uv_executor_done_func uv_executor_t.done
+
+    Assigned by libuv.
+    Executor should invoke this when work is done or successfully
+    cancelled.
+
+.. c:member:: void * uv_executor_t.data
+
+    Space for user-defined arbitrary data. libuv does not use this field.
+
+.. c:member:: uv_work_type uv_work_options_t.type
+
+    Type of request. Readonly.
+
+    ::
+
+        typedef enum {
+            UV_WORK_UNKNOWN = 0,
+            UV_WORK_FS,
+            UV_WORK_DNS,
+            UV_WORK_USER_IO,
+            UV_WORK_USER_CPU,
+            UV_WORK_PRIVATE,
+            UV_WORK_MAX = 255
+        } uv_work_type;
+
+.. c:member:: int uv_work_options_t.priority
+
+    Suggested for use in user-defined executor.
+    Has no effect on libuv's default executor.
+
+.. c:member:: int uv_work_options_t.cancelable
```

Comment:
Correct, it is not currently used. This would permit the executor to attempt to forcibly abort tasks that are already running on a worker. Since `pthread_cancel` is pretty dangerous, I'm not sure we want to go there, but I wanted to float the idea.

As an alternative I suppose the executor could detect long-running work and spin up replacement workers instead.

If the long-running work is hung on I/O (e.g. a blocking read), pthread_cancel probably won't kill it until the I/O finishes. So this would mostly be helpful for "stuck" CPU-bound work that will continue to burn the CPU until it finishes or it is aborted.